docs/writing-rules/glossary.md

```diff
@@ -0,0 +1,124 @@
+
+## Error matrix
+
+An error matrix is a 2x2 table that visualizes the performance of a Semgrep rule in relation to the findings it does or doesn't detect. It has two axes:
+
+- Positive and negative
+- True or false
+
+These yield the following combinations:
+
+<dl>
+<dt>True positive</dt>
+<dd>The rule detected a piece of code it was intended to find.</dd>
+<dt>False positive</dt>
+<dd>The rule detected a piece of code it was not intended to find.</dd>
+<dt>True negative</dt>
+<dd>The rule correctly skipped over a piece of code it wasn't meant to find.</dd>
+<dt>False negative</dt>
+<dd>The rule failed to detect a piece of code it should have found.</dd>
+</dl>
+
+Not to be confused with **risk matrices**.
+
+## Finding
+
+A finding is the core result of Semgrep's analysis. Findings are generated when a Semgrep rule matches a piece of code. Findings can be security issues, bugs, or code that doesn't follow coding conventions.
+
+## Fully-qualified name
+
+A fully-qualified name refers to a name which uniquely identifies a class, method, type, or module. Languages such as C# and Ruby use `::` to distinguish between fully-qualified names and regular names.
+
+Not to be confused with **tokens**.
+
+## l-value (left-, or location-value)
+
+An expression that denotes an object in memory; a memory location, something that you can use in the left-hand side (LHS) of an assignment.
+
+Compare to [r-value](https://learn.microsoft.com/en-us/cpp/c-language/l-value-and-r-value-expressions?view=msvc-170).
+
+## Metavariable
+
+A metavariable is an abstraction that lets you match something even when you don't know exactly what it is you want to match. It is similar to capture groups in regular expressions. All metavariables begin with a `$` and can only contain uppercase characters, digits, and underscores.
+
+## Propagator
+
+A propagator is any code that alters a piece of data as the data moves across the program. This includes functions, reassignments, and so on.
+
+## Rule (Semgrep rule)
+
+A rule is a specification of the patterns that Semgrep must match to the code to generate a finding. Rules are written in YAML. Without a rule, the engine has no instructions on how to match code.
+
+Rules can be run on either Semgrep OSS Engine or Semgrep Pro Engine.
+
+There are two types of rules: search and taint.
+
+<dl>
+<dt>Search rules</dt>
+<dd>Rules default to this type. Search rules simply detect matches based on the patterns described by a rule.</dd>
+<dt>Taint rules</dt>
+<dd>Taint rules make use of Semgrep's taint analysis in addition to default search functionalities. Taint rules are able to specify sources, sinks, and propagators of data as well as sanitizers of that data. For more information, see Taint analysis documentation (tk link)</dd>
+</dl>
+
+<!-- how can we say that search rules are semantic if no analysis is performed on the value of data, such as variables? Or are there levels of semantic understanding that semgrep can perform? -->
+
```

Comment:
Here are a few things we do for search mode rules:
1. During the parsing and matching phases, we might interpret syntactically different code as semantically the same. (I can't think of a good example right now but you could imagine that we could parse for loops and while loops into the same ast type, and then for loops would match while loops)
2. If you use the fully qualified name of a function in the pattern, we will match that function even when it isn't referenced with the fully qualified name
3. We perform constant propagation (it's really just constant folding, we don't do anything fancy, but it's still semantic)
4. We have typed metavariables, which require some amount of type inference

Reply:
I added these in! Thank you. 

![image](https://github.com/semgrep/semgrep-docs/assets/20766840/c2bc858c-1dd7-454e-a33a-0a46a664a689)